src/GraphQL/DI/IGraphQLBuilder.cs

```diff
@@ -0,0 +1,41 @@
+using System;
+
+namespace GraphQL.DI
+{
+    /// <summary>
+    /// An interface for configuring GraphQL.NET services.
+    /// </summary>
+    public interface IGraphQLBuilder
+    {
+        /// <summary>
+        /// Registers the service of type <paramref name="serviceType"/> with the dependency injection provider.
+        /// </summary>
+        IGraphQLBuilder Register(Type serviceType, Type implementationType, ServiceLifetime serviceLifetime);
+
+        /// <inheritdoc cref="Register(Type, Type, ServiceLifetime)"/>
+        IGraphQLBuilder Register(Type serviceType, Func<IServiceProvider, object> implementationFactory, ServiceLifetime serviceLifetime);
+
+        /// <inheritdoc cref="Register(Type, Type, ServiceLifetime)"/>
+        IGraphQLBuilder Register(Type serviceType, object implementationInstance);
+
+        /// <summary>
+        /// Registers the service of type <paramref name="serviceType"/> with the dependency injection provider if a service
+        /// of the same type has not already been registered.
+        /// </summary>
+        IGraphQLBuilder TryRegister(Type serviceType, Type implementationType, ServiceLifetime serviceLifetime);
+
+        /// <inheritdoc cref="TryRegister(Type, Type, ServiceLifetime)"/>
+        IGraphQLBuilder TryRegister(Type serviceType, Func<IServiceProvider, object> implementationFactory, ServiceLifetime serviceLifetime);
+
+        /// <inheritdoc cref="TryRegister(Type, Type, ServiceLifetime)"/>
+        IGraphQLBuilder TryRegister(Type serviceType, object implementationInstance);
+
+        /// <summary>
+        /// Configures an options class of type <typeparamref name="TOptions"/>.
+        /// <br/><br/>
+        /// Passing <see langword="null"/> as the delegate is allowed and will skip this registration.
```

Comment:
> is allowed and **will skip** this registration

Comment confuses.
![изображение](https://user-images.githubusercontent.com/21261007/147165066-a653f560-3acf-462c-b754-4f3923293c02.png)

Actually registration is not skipped. 1 - registration, 2 - additional registration (or configuration) .

Such comment is a consequence of leaky abstraction.

Reply:
I don’t understand what you mean by “leaky abstraction”. 

The comment could be written better. It should say that if action is null then TOptions is still configured and will return a default instance (unless otherwise configured with a subsequent call to Configure). 

This design is intended and required. All calls to Configure execute for the same TOptions, if you know what I mean.

Reply:
> It should say that if action is null then TOptions is still configured and will return a default instance (unless otherwise configured with a subsequent call to Configure).

Exactly.

Reply:
#2781